core/kernel.go

```diff
@@ -112,6 +112,7 @@ func NewKernel(ctx context.Context, keyClient keys.KeyClient, privValidator tmTy
 	} else {
 		kern.State, err = execution.MakeGenesisState(stateDB, genesisDoc)
 	}
+	kern.Blockchain.BlockHashProvider = kern.State.GetBlockHash
```

Comment:
Hm I'm not sure how I feel about patching it in like this. In the short term its nice and simple but it creates an ad hoc coupling between state and blockchain that doesn't pass through the standard constructor or other interfaces. I feel like it degrades our separate a bit. Let me reflect.